src/content/contribute/writing-a-loader.md

```diff
@@ -48,93 +39,111 @@ resolveLoader: {
+## Guidelines
 
-Loader generated modules should respect the same design principles like normal modules.
+The following guidelines should be followed when writing a loader. They are ordered in terms of importance and some only apply in certain scenarios, read the detailed sections that follow for more information.
 
-Example: That's a bad design: (not modular, global state, ...)
+- Keep them __simple__.
+- Utilize __chaining__.
+- Emit __modular__ output.
+- Make sure they're __stateless__.
+- Employ __loader utilities__.
+- Mark __loader dependencies__.
+- Resolve __module dependencies__.
+- Extract __common code__.
+- Avoid __absolute paths__.
+- Use __peer dependencies__.
 
-```javascript
-require("any-template-language-loader!./xyz.atl");
+### Simple
 
-var html = anyTemplateLanguage.render("xyz");
-```
+Loaders should do only a single task. This not only makes the job of maintaining each loader easier, but also allows them to be chained for usage in more scenarios.
+
+### Chaining
+
+Take advantage of the fact that loaders can be chained together. Instead of writing a single loader that tackles five tasks, write five simpler loaders that divide this effort. Isolating them not only keeps each individual loader simple, but may allow for them to be used for something you hadn't though of originally.
+
+Take the case of rendering a template file with data specified via loader options or query parameters. It could be written as a single loader that compiles the template from source, executes it and returns a module that exports a string containing the HTML code. However, in accordance with guidelines, a simple `apply-loader` exists that can be chained with other open source loaders:
 
-### Do not keep state between runs and modules
+- `jade-loader`: Convert template to a module that exports a function.
+- `apply-loader`: Executes the function with loader options and returns raw HTML.
+- `html-loader`: Accepts HTML and outputs a valid JavaScript module.
 
-A loader should be independent of other modules compiled (except of these issued by the loader).
+T> The fact that loaders can be chained also means they don't necessarily have to output JavaScript. As long as the next loader in the chain can handle its output, the loader can return any type of module.
 
-A loader should be independent of previous compilations of the same module.
+### Modular
 
-### Use the [loader-utils](https://github.com/webpack/loader-utils)
+Keep the output modular. Loader generated modules should respect the same design principles as normal modules.
 
-In order to make the experience consistent for other developers, you should use the loader-utils to get the loader options:
+### Stateless
 
-```javascript
-const loaderUtils = require("loader-utils");
+Make sure the loader does not retain state between module transformations. Each run should always be independent of other compiled modules as well as previous compilations of the same module.
+
+### Loader Utilities
```

Comment:
```
## Utilities

### Loader
...
### Schema
...

Reply:
Same reason as above plus these headers line up with the bullets outlined at the beginning of the section.